keps/sig-network/3458-dynamically-re-configure-lb-node-set/README.md

```diff
@@ -0,0 +1,787 @@
+
+This section describes how Kube-proxy handles all above cases currently.
+
+A) Kube-proxy does not handle the first case. If Kube-proxy restarts it will stop
+  answering to the health check probes while it's not running, this is because
+  the health check server runs as a go-routine in the Kube-proxy process.  
+
+B) Kube-proxy currently validates that whatever service/endpoints event it
+  receives are also processed within a specific timeout, which is configurable.
+  Any event received will set the field `oldestPendingQueued` on its health
+  check server implementation. The value will correspond to the time the event
+  was observed and once that event is processed successfully: the same field is
+  unset to a zero time value. The health check server performs two validations
+  and reports green if: that field is the zero value, or if the field is not the
+  zero value but within the timeout range. This currently handles the "general
+  API server read issue" case mentioned above, whereby if all Kube-proxies
+  encounter issues reading from the API server: the state of the healthz
+  endpoints will indicate green from all of them, given that the last event
+  received was processed successfully. It does not handle the case where one / a
+  group of Kube-proxy replicas can't read from the API server, since these
+  replicas will continue reporting green, but probably should not be. It could
+  be that Kube-proxy is using invalid credentials when communicating with the
+  API server and that the node is not actually experiencing a loss of API server
+  connectivity. This case could have pods move around without Kube-proxy being
+  able to update its rules. In the worse case: this could have Kube-proxy point
+  load balancer traffic to old pod IPs which have been recycled and are now
+  being used by new pods - leading to security concerns.
+
+C) The third case is partially covered by the validation performed Kube-proxy
+  [when it initializes](1) - which validates that the configuration for its
+  dependent networking sub-system is correctly configured. Its "runtime" checks
+  are covered by it's current health check, since, if it experiences any issues
+  updating the networking rules: it won't set `oldestPendingQueued` to the zero
+  time value and hence fail the health check.
+
+It is important to mention that all of these checks are currently being
+performed for class A) services only, and should probably be extended to class
+B) as well. The health check for class B) services should hence answer if
+endpoints are running on the node AND if Kube-proxy is healthy.
+
+[1]: https://github.com/kubernetes/kubernetes/blob/f9afd68e3b3fc05f3607ea7cb90808c3e931ba9c/pkg/proxy/apis/config/validation/validation.go#L38
+
+#### Defining what "healthy" should be for Kube-proxy
+
+This section describes how Kube-proxy might handle covering all above cases.
+
+As mentioned above, A) is not handled by Kube-proxy today.
```

Comment:
The only way I can think to solve this is to split processes.  E.g at startup, `fork()` and make the child process run kube-proxy while the parent serves HC.  If kube-proxy crashes, we would have to restart it.  Unfortunate to duplicate so much logic.

Alternately, run a small HTTP server process that reads state from a shared volume, and makw kube-proxy write that state.  This requires more work for people deploying k-p but gets extra resilience.  

We could make this optional, like `kube-proxy --run-hc-server=<path>` means to become the server, and `--use-hc-server=<path>` means to write data to the volume.  Default means "serve HC internally".

This could be a KEP of its own.

Reply:
>The only way I can think to solve this is to split processes. E.g at startup, fork() and make the child process run kube-proxy while the parent serves HC. If kube-proxy crashes, we would have to restart it. Unfortunate to duplicate so much logic.

Kube-proxy runs as a Pod today. If we just `fork` a child process from Kube-proxy, that process will still run in the Pod's namespace and be impacted by any restart that the Pod experiences, right? It seems to me like we'd need something `fork`'ed in the host's PID namespace.  

> Alternately, run a small HTTP server process that reads state from a shared volume, and makw kube-proxy write that state. This requires more work for people deploying k-p but gets extra resilience.

That might work if the Kubelet runs this HTTP server. But then it might as well just run the health check server altogether.

> This could be a KEP of its own.

Agreed.  There's also been a bit of discussion in the sig-net mailing list around "should the CRI include networking". This has no link to CNI/CRI, but maybe might be a topic of conversation for the longer vision of Kube networking?

Reply:
>> Alternately, run a small HTTP server process that reads state from a shared volume, and makw kube-proxy write that state. This requires more work for people deploying k-p but gets extra resilience.
>
> That might work if the Kubelet runs this HTTP server. But then it might as well just run the health check server altogether.

You don't need kubelet to run it, you just need a different container.  IOW, kube-proxy pods would run 2 containers, one for the HC server, one for the "proxy" (it's not a proxy), and commuinicate over a shared emptyDir volume (probably with medium = Memory).  If k-p fails, we restart it.  If the whole pod is restarted, then yeah, HCs would fail.

Reply:
>  If the whole pod is restarted, then yeah, HCs would fail.

But that's what I would hope we fix. 

Upgrading the kube-proxy pod shouldn't impact ingress. On the other hand: if kube-proxy is failing, then I think it should, because we can't be sure it's been able to complete its work successfully.

Reply:
I'm looking for baby steps.  If we can break it to a different container, then we can break it to a different pod or systemd service or whatever.

First we need to define what "healthy" means and how to expose that information out-of-process.

Reply:
Punted to a later KEP